x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/MlNodeShutdownIT.java

```diff
@@ -100,6 +101,97 @@ public void testJobsVacateShuttingDownNode() throws Exception {
         });
     }
 
+    public void testCloseJobVacatingShuttingDownNode() throws Exception {
+
+        // TODO: delete this condition when the shutdown API is always available
+        assumeTrue("shutdown API is behind a snapshot-only feature flag", Build.CURRENT.isSnapshot());
+
+        internalCluster().ensureAtLeastNumDataNodes(3);
+        ensureStableCluster();
+
+        // index some datafeed data
+        client().admin().indices().prepareCreate("data")
+            .setMapping("time", "type=date")
+            .get();
+        long numDocs1 = randomIntBetween(50, 100);
+        long now = System.currentTimeMillis();
+        long weekAgo = now - 604800000;
+        long twoWeeksAgo = weekAgo - 604800000;
+        indexDocs(logger, "data", numDocs1, twoWeeksAgo, weekAgo);
+
+        // Open 6 jobs.  Since there are 3 nodes in the cluster we should get 2 jobs per node.
+
+        setupJobAndDatafeed("shutdown-close-job-1", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-2", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-3", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-4", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-5", ByteSizeValue.ofMb(2));
+        setupJobAndDatafeed("shutdown-close-job-6", ByteSizeValue.ofMb(2));
+
+        // Call the shutdown API for one non-master node.
+        String nodeNameToShutdown = Arrays.stream(internalCluster().getNodeNames())
+            .filter(nodeName -> internalCluster().getMasterName().equals(nodeName) == false).findFirst().get();
+        SetOnce<String> nodeIdToShutdown = new SetOnce<>();
+        SetOnce<String> jobIdToClose = new SetOnce<>();
+
+        assertBusy(() -> {
+            GetJobsStatsAction.Response statsResponse =
+                client().execute(GetJobsStatsAction.INSTANCE, new GetJobsStatsAction.Request(Metadata.ALL)).actionGet();
+            QueryPage<GetJobsStatsAction.Response.JobStats> jobStats = statsResponse.getResponse();
+            assertThat(jobStats, notNullValue());
+            long numJobsOnNodeToShutdown = jobStats.results().stream()
+                .filter(stats -> stats.getNode() != null && nodeNameToShutdown.equals(stats.getNode().getName())).count();
+            long numJobsOnOtherNodes = jobStats.results().stream()
+                .filter(stats -> stats.getNode() != null && nodeNameToShutdown.equals(stats.getNode().getName()) == false).count();
+            assertThat(numJobsOnNodeToShutdown, is(2L));
+            assertThat(numJobsOnOtherNodes, is(4L));
```

Comment:
Is it conceivable that things move SO fast that the jobs all get unassigned too quick so that this assertion always fails. I think assertBusy has an exponential backoff, so it seems to me that this could cause the test to be flaky.

Reply:
The shutdown API is called after this `assertBusy` block, not before.  I think my comments might have been confusing because I had a comment next to the declaration of the variables that remember which node is being shut down saying "Call the shutdown API" so I have made the comments more accurate about which bits do what.

There's a complication because the internal cluster test functionality works in terms of node names and the shutdown API works in terms of node IDs, so we have to find the node ID corresponding to a node name.  Since the ML stats API coincidentally returns that information for nodes where ML jobs are running I got it from the output of that call rather than making a separate call.